### PR TITLE
[Dust apps] UI cleanup + UX : add possibility to live-edit the dataset's data from the input block

### PR DESCRIPTION
**Pull request by albandum**
## Description

Building dust apps is complicated when you first get to it because it's unclear that the data of your input dataset is used when clicking on the "RUN" button (as a test data) 
## UX
- Made it possible to view and directly edit the dataset's data in the INPUT block 
- Update through a modal view
Screenshot 2024-05-02 at 12 48 05

## UI
- Can still switch between datasets 
- Changed the blocks dropdown to show block name instead of block type (type is still show when the block actually is created)
- Removed LLM block in input choice as we are deprecating in favor of chat 
- Added tooltip on cache icon to make it a bit understandable

Screenshot 2024-04-26 at 17 47 13
Screenshot 2024-04-26 at 17 53 52

## Risk

As in the entire dust app builder, if you screw up it's auto saved without a history 

## Deploy Plan
No migration

### CODE REVIEW

File: front/components/app/NewBlock.tsx [hunk 1]
```diff
@@ -166,14 +166,13 @@ export default function NewBlock({
...
-                      >
-                        {type}
-                      </div>
-                    ))}
+                    <div
+                      className={`mr-1 rounded-md px-1 py-0.5 text-sm font-bold ${
+                        block.type === "input" ? "bg-orange-200" : "bg-gray-200"
+                      }`}
+                    >
+                      {block.name}
```
Comment (albandum) [inline front/components/app/NewBlock.tsx:174]: Kind of expecting @spolu to say that we still need to display the type here. Question is why? Type is likely more important in the docs that in the webapp?

File: front/components/app/blocks/Block.tsx [hunk 2]
```diff
@@ -129,23 +130,27 @@ export default function Block({
...
             ) : (
-              <div
-                className="flex flex-initial cursor-pointer text-gray-400"
-                onClick={() => {
-                  handleUseCacheChange(true);
-                }}
-              >
-                <ArrowPathIcon className="h-4 w-4" />
-              </div>
+              <Tooltip label="Results are not cached (slower)" position="above">
```
[the remark this answers is not in the document]
Comment (spolu) [reply front/components/app/blocks/Block.tsx:144]: Maybe `Results are recomputed at each run` ?

File: front/components/app/NewBlock.tsx [hunk 3]
```diff
@@ -27,16 +27,10 @@ export default function NewBlock({
     name: string;
     description: string;
   }[] = [
-    {
```
Comment (spolu) [inline front/components/app/NewBlock.tsx:30]: Bullish to remove it but that means removing it from the docs as well. This should be in a separate PR IMHO

File: front/components/app/NewBlock.tsx (hunk 1 above)
Comment (spolu) [reply front/components/app/NewBlock.tsx:174]: Let's keep type for now as it is i,portant that these tag match the type tag on each block :+1: 

We can revisit that later of course.

File: front/components/app/blocks/InputBlock.tsx [hunk 4]
```diff
@@ -0,0 +1,167 @@
+import { Button, PencilSquareIcon } from "@dust-tt/sparkle";
```
Comment (spolu) [inline front/components/app/blocks/InputBlock.tsx:1]: Why rename? this breaks the nomemclature. Let's keep the file name and block name ? Will make the PR easier to review as well.

File: front/components/app/DatasetPicker.tsx [hunk 5]
```diff
@@ -69,16 +62,17 @@ export default function DatasetPicker({
...
+                "inline-flex items-center rounded-md px-3 py-1 text-sm font-normal",
+                dataset ? "border px-1" : "border border-orange-400",
+                readOnly ? "border-white text-gray-300" : "text-gray-700",
                 "focus:outline-none focus:ring-0"
               )}
             >
               {dataset ? (
                 <>
+                  <div className="text-sm font-bold text-action-500">
+                    {dataset}
```
Comment (spolu) [inline front/components/app/DatasetPicker.tsx:74]: How can one edit the dataset from that block? Is that intended that it is not possible anymore? This is OK just want to check it's intentional.

File: front/pages/w/[wId]/a/index.tsx [hunk 6]
```diff
@@ -566,7 +566,7 @@ export default function Developers({
   const router = useRouter();
   const handleTabChange = async (tabId: string) => {
     const query = { ...router.query, t: tabId };
-    await router.push({ query });
+    await router.replace({ query });
```
Comment (spolu) [inline front/pages/w/[wId]/a/index.tsx:569]: Why replace?

**Review by spolu — CHANGES_REQUESTED** [review]
Left a first batch of questions + change requests :+1: 

Haven't reviewed the new Input block closely yet because the file rename makes it very hard to review.

File: front/pages/w/[wId]/a/index.tsx (hunk 6 above)
Comment (albandum) [reply front/pages/w/[wId]/a/index.tsx:569]: For cleaner browser history management - cf https://github.com/dust-tt/dust/pull/4845#pullrequestreview-2022712693

File: front/components/app/DatasetPicker.tsx (hunk 5 above)
Comment (albandum) [reply front/components/app/DatasetPicker.tsx:74]: It felt quite counterintuitive to click on the dropdown and be redirected to editing the dataset - never seen that in a UI 
So I just added an edit button next to it 
Screenshot 2024-04-29 at 10 22 49

File: front/components/app/NewBlock.tsx (hunk 3 above)
Comment (albandum) [reply front/components/app/NewBlock.tsx:30]: Make sense, will do

File: front/components/app/NewBlock.tsx (hunk 1 above)
Comment (albandum) [reply front/components/app/NewBlock.tsx]: Screenshot 2024-04-29 at 10 35 21
Does this look OK?

File: front/components/app/DatasetPicker.tsx (hunk 5 above)
Comment (spolu) [reply front/components/app/DatasetPicker.tsx:74]: No a strong reason enough :)

Comment (spolu) [reply front/components/app/DatasetPicker.tsx:74]: Well these tabs are "more engaging" IMHO. And I personally do rely on browser history to move from app to datasets to config etc...

Comment (albandum) [reply front/components/app/DatasetPicker.tsx:74]: > No a strong reason enough :)

I personally never click on the down arrow to open a dropdown and I don't expect clicking in the middle to be link. 
Having a counterintuitive dropdown behavior seems like a quite strong reason to make a small change to me haha, but I can revert if that doesn't seem right. 🥲

Comment (albandum) [reply front/components/app/DatasetPicker.tsx:74]: > Well these tabs are "more engaging" IMHO. And I personally do rely on browser history to move from app to datasets to config etc...

I believe this was meant for the `router.replace` part - reverted, we can decide later.

File: front/pages/w/[wId]/a/index.tsx (hunk 6 above)
Comment (spolu) [reply front/pages/w/[wId]/a/index.tsx:569]: Well these tabs are "more engaging" IMHO. And I personally do rely on browser history to move from app to datasets to config etc...

File: front/components/app/blocks/InputBlock.tsx (hunk 4 above)
Comment (spolu) [reply front/components/app/blocks/InputBlock.tsx]: No a strong reason enough :)

File: front/components/app/DatasetPicker.tsx (hunk 5 above)
Comment (spolu) [reply front/components/app/DatasetPicker.tsx:74]: Comments were not in the right place. This change looks good to me :+1:

File: front/components/app/NewBlock.tsx [hunk 7]
```diff
@@ -30,48 +30,48 @@ export default function NewBlock({
...
-      name: "DataSource",
+      name: "Search a datasource",
       description:
         "Perform semantic search against chunked documents from a DataSource.",
     },
     {
       type: "curl",
       typeNames: ["curl"],
-      name: "Curl Request",
+      name: "CURL Request",
```
Comment (spolu) [inline front/components/app/NewBlock.tsx:67]: To be super geek let's do cURL

**Review by spolu — COMMENTED** [review]
Small comments (continuing review)

File: front/pages/api/w/[wId]/apps/[aId]/datasets/[name]/index.ts [hunk 8]
```diff
@@ -62,24 +63,40 @@ async function handler(
...
+    });
+  }
+
+  if (!auth.isBuilder()) {
+    return apiError(req, res, {
+      status_code: 403,
+      api_error: {
+        type: "app_auth_error",
+        message:
+          "Only the users that are `builders` for the current workspace interact with datasets",
```
Comment (spolu) [inline front/pages/api/w/[wId]/apps/[aId]/datasets/[name]/index.ts:78]: missing a can?

File: front/components/app/blocks/Input.tsx [hunk 9]
```diff
@@ -40,12 +45,67 @@ export default function Input({
...
+  );
+
+  const handleSetDataset = async (dataset: string) => {
     const b = shallowBlockClone(block);
     b.config.dataset = dataset;
+    setDatasetWithData(await handleGetDatasetData(dataset));
     onBlockUpdate(b);
   };
 
+  const handleGetDatasetData = async (dataset: string) => {
```
Comment (spolu) [inline front/components/app/blocks/Input.tsx:59]: This should be an SWR no?

File: front/components/app/blocks/Input.tsx [hunk 10]
```diff
@@ -40,12 +45,67 @@ export default function Input({
...
 }>) {
-  const handleSetDataset = (dataset: string) => {
+  const [datasetWithData, setDatasetWithData] = useState<DatasetType | null>(
+    null
+  );
+
+  const handleSetDataset = async (dataset: string) => {
     const b = shallowBlockClone(block);
     b.config.dataset = dataset;
+    setDatasetWithData(await handleGetDatasetData(dataset));
```
Comment (spolu) [inline front/components/app/blocks/Input.tsx:55]: This is a very slow call that we don't in the logic of adding/editing the Input block no?

Shuldn't this be an SWR hook that depends on b.config.dataset  ?

File: front/components/app/blocks/Input.tsx [hunk 11]
```diff
@@ -63,21 +123,46 @@ export default function Input({
...
+                  icon={PencilSquareIcon}
+                  label="Edit schema"
+                  size="xs"
+                />
+              ) : null}
+            </div>
           ) : null}
         </div>
+
+        {datasetWithData && datasetWithData.schema ? (
```
Comment (spolu) [inline front/components/app/blocks/Input.tsx:153]: The dataset does not render full width which feels super weird?

File: front/components/app/blocks/Input.tsx [hunk 12]
```diff
@@ -63,21 +123,46 @@ export default function Input({
...
+          <div className="max-h-[800px] overflow-y-auto">
```
Comment (spolu) [inline front/components/app/blocks/Input.tsx:154]: I'm not a massive fan of the double scrolling here?

Comment (spolu) [reply front/components/app/blocks/Input.tsx:154]: I'm still not 100% convinced by this :) Let's discuss more :+1:

**Review by spolu — COMMENTED** [review]
Left a few more comments. I think we can improve the gestion of retrieving the dataset to display in the input block

File: front/pages/api/w/[wId]/apps/[aId]/datasets/[name]/index.ts (hunk 8 above)
Comment (albandum) [reply front/pages/api/w/[wId]/apps/[aId]/datasets/[name]/index.ts:78]: 🥫fixed

File: front/components/app/DatasetView.tsx [hunk 13]
```diff
@@ -423,171 +425,203 @@ export default function DatasetView({
     datasetTypes,
   ]);
 
+  useEffect(() => {
```
[the remark this answers is not in the document]
Comment (albandum) [reply front/components/app/DatasetView.tsx:428]: That's keeping the state up-to-date after a dataset switch - I think it could be avoided using SWR but in all of dust apps it looks like we're not using SWR hooks, so I kept the same style (I believe?)

[the remark this answers is not in the document]
Comment (spolu) [reply front/components/app/DatasetView.tsx:433]: Yep but it blocks the main rendering loop here the block appearance can become extremely slow

Agreed there is no SWR in Dust apps because it's the oldest part of the code but I thiiink this is what we want as we update this behavior?

File: front/components/app/NewBlock.tsx [hunk 14]
```diff
@@ -166,17 +173,18 @@ export default function NewBlock({
...
-                    {block.typeNames.map((type) => (
-                      <div
-                        key={type}
-                        className="mr-1 rounded-md bg-gray-200 px-1 py-0.5 text-sm font-bold"
-                      >
-                        {type}
-                      </div>
-                    ))}
+                    <div
+                      className={`mr-1 rounded-md px-1 py-0.5 text-sm font-bold ${
```
[the remark this answers is not in the document]
Comment (albandum) [reply front/components/app/NewBlock.tsx:176]: Not sure why you're saying so? I don't see anything broken on my end (or I'm not looking at the right place?)


Screenshot 2024-04-29 at 14 26 22


Screenshot 2024-04-29 at 14 25 58

File: front/components/app/DatasetView.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (albandum) [reply front/components/app/DatasetView.tsx:433]: Ok ! I'll need to dive in more into SWR then, will look 👀

File: front/components/app/NewBlock.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (spolu) [reply front/components/app/NewBlock.tsx:176]: This guy no? `map` `reduce` as an example here

Screenshot from 2024-04-29 16-30-40

File: front/components/app/blocks/Input.tsx (hunk 12 above)
Comment (albandum) [reply front/components/app/blocks/Input.tsx]: Last try with one quick change -> modal on the side. 2 clicks to go to schema edition but this fixes the 'too long' issue and allows iterating. Wdyt @spolu ?

https://github.com/dust-tt/dust/assets/1189312/e892697c-3d8a-44a8-b0db-7ecbdf5d90e2

**Comment by spolu** [comment]
As per IRL

Let's add a button for edition + other small improvements in that PR but let's not do inline edition.

Instead let's spend some time having an input block with a dataset loaded at app creation with a better (less confusing) default dataset.

File: front/components/app/NewBlock.tsx [hunk 15]
```diff
@@ -27,6 +27,12 @@ export default function NewBlock({
     name: string;
     description: string;
   }[] = [
+    {
```
Comment (spolu) [inline front/components/app/NewBlock.tsx:30]: Beware you're re-adding it here

File: front/components/app/blocks/Input.tsx [hunk 16]
```diff
@@ -40,12 +46,43 @@ export default function Input({
...
+    useDataset(owner, app, block.config.dataset, true);
+
+  const handleSetDataset = async (datasetName: string) => {
     const b = shallowBlockClone(block);
-    b.config.dataset = dataset;
+    b.config.dataset = datasetName;
     onBlockUpdate(b);
   };
 
+  const onUpdate = _.debounce(
```
Comment (spolu) [inline front/components/app/blocks/Input.tsx:59]: As per IRL I think this may overwrite the state of your data set if the debounce + write + get races with a user edit? To be confirmed

Proposal as per IRL is to have a save/cancel (+ confirm in case of exsit by click) on the modal.

File: front/components/app/blocks/Data.tsx [hunk 17]
```diff
@@ -82,6 +83,17 @@ export default function Data({
...
             />
           )}
+          {block.spec.dataset && (
+            <Button
+              variant="secondary"
+              onClick={() => {
+                window.location.href = `/w/${owner.sId}/a/${app.sId}/datasets/${block.spec.dataset}`;
+              }}
+              icon={PencilSquareIcon}
+              label="Edit"
```
Comment (spolu) [inline front/components/app/blocks/Data.tsx:93]: Do we want the modal here as well?

**Comment by spolu** [comment]
(We're getting there thanks for bearing with me :p)

File: front/components/app/blocks/Data.tsx (hunk 17 above)
Comment (albandum) [reply front/components/app/blocks/Data.tsx:93]: Wdyt? I think it's less necessary - `data` block seems less likely to need quick iteration that `input` block, but we can add as well

Comment (spolu) [reply front/components/app/blocks/Data.tsx:93]: Just create two different experience for something that look exactly the same?

Comment (albandum) [reply front/components/app/blocks/Data.tsx:93]: What do we do about this paragraph then? Do I just add a flag to remove it?

Screenshot 2024-05-02 at 12 44 23

File: front/components/app/DatasetView.tsx [hunk 18]
```diff
@@ -423,171 +430,203 @@ export default function DatasetView({
...
+    // If in block view, update the dataset keys and data
+    if (dataset && viewType == "block") {
```
Comment (spolu) [inline front/components/app/DatasetView.tsx:435]: Do we really still need that?

File: front/components/app/blocks/Input.tsx [hunk 19]
```diff
@@ -40,12 +46,62 @@ export default function Input({
...
+    block.config.dataset,
+    true
+  );
+
+  const [datasetModalData, setDatasetModalData] = useState<DatasetType | null>(
+    null
+  );
+  const sendNotification = useContext(SendNotificationsContext);
+
+  const handleSetDataset = async (datasetName: string) => {
```
Comment (spolu) [inline front/components/app/blocks/Input.tsx:62]: can we keep `dataset`? this is indeed the name but this is also an `id` and using `datasetName` really does not convey that :pray:

File: front/components/app/blocks/Input.tsx [hunk 20]
```diff
@@ -40,12 +46,62 @@ export default function Input({
...
     const b = shallowBlockClone(block);
-    b.config.dataset = dataset;
+    b.config.dataset = datasetName;
     onBlockUpdate(b);
   };
 
+  const onUpdate = (
```
Comment (spolu) [inline front/components/app/blocks/Input.tsx:68]: Let's inline this one in the modal code directly :pray:

**Review by spolu — COMMENTED** [review]
Generally LGTM Left a few comments :+1:

File: front/components/app/DatasetView.tsx (hunk 18 above)
Comment (albandum) [reply front/components/app/DatasetView.tsx:435]: we don't - good catch

**Review by spolu — APPROVED** [review]
LGTM \o/
